### PR TITLE
[client] Add support for protocol (ray://, local://, custom://) to ray.init

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

Step towards a single API to start ray (instead of the existing ray.init vs ray.client/ray.util.connect). 

ray.init now works as follows:
If address=None or address="someString", then behaves **identically** to current behavior.

If address="someProtocol://someString", then use the ClientBuilder from someProtocol to start a client session. This will default to the regular client builder for "ray://", and to _LocalClientBuilder for "local://"

ray.init now takes catchall **kwargs. If there are any unused **kwargs an error is thrown, to make sure typos get caught in a noisy way, so that custom ClientBuilders can take extra arguments.

All arguments to init are passed to ClientBuilder._init_args, which custom builders can override to handle new parameters and error on invalid params.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by WangTaoTheTonic** [comment]
I'm not sure if we should go further by using `ray.init` to start a ray cluster.
From what I see `ray.init` is used for initializing a ray context. Only in local env will it start a cluster when there's no existing one, for convenience, is it right?

File: python/ray/worker.py
Comment (AmeerHajAli) [inline python/ray/worker.py:721]: why do we need this?
I think I am missing something.

**Comment by AmeerHajAli** [comment]
Considering the caveats of the API, I would want to make sure @ijrsvt signs off on this.

File: python/ray/client_builder.py
Comment (AmeerHajAli) [inline python/ray/client_builder.py:85]: why do we need to add dummyProto?

File: python/ray/worker.py
Comment (ckw017) [reply python/ray/worker.py:721]: The check for "://" in address is needed since the current plan is to make sure that init's behavior remains the same when the address doesn't have a protocol, which should keep us from breaking any existing code (existing code shouldn't have "someProtocol://" in the address, afaik)

The use of ray.client here is a bit conflicting since technically we would want to phase out ray.client, but it's currently where all the plugin logic (like loading a custom ClientBuilder) lives, and since we'll be keeping ClientBuilder working I figured it would be useful to just reuse that logic for now

File: python/ray/client_builder.py
Comment (AmeerHajAli) [inline python/ray/client_builder.py:95]: We are only using VARIABLES per argument, so no URL arguments.

File: python/ray/tests/test_ray_init.py
Comment (AmeerHajAli) [inline python/ray/tests/test_ray_init.py:150]: I don't remember seeeing "local://" in the API doc.

File: python/ray/client_builder.py
Comment (ckw017) [reply python/ray/client_builder.py:85]: If the address is something like "localhost:50005?query=thing" then urlparse will mix up `localhost` as the scheme (protocol)

```python
>>> urlparse("localhost:8000?a=b")
ParseResult(scheme='localhost', netloc='', path='8000', params='', query='a=b', fragment='')
>>> urlparse("hi://localhost:8000?a=b")
ParseResult(scheme='hi', netloc='localhost:8000', path='', params='', query='a=b', fragment='')
```

^ in the second line specifying the scheme makes sure the whole address lands in `netloc`

**Comment by AmeerHajAli** [comment]
We also need to make the documentation changes to only reflect ray.init() API and remove the client builder API from the documentation in OSS and other places.

File: python/ray/client_builder.py
Comment (AmeerHajAli) [reply python/ray/client_builder.py:85]: but wait, this looks like an illegal address, because we only support
XXX://<string>
or
<string>
you can't do localhost:50005?query=thing. Actually you can't add `?query=thing` to any option
.

File: python/ray/worker.py
Comment (AmeerHajAli) [reply python/ray/worker.py:721]: maybe a more proper way is to say:
`builder.store_args()`?

File: python/ray/worker.py
Comment (AmeerHajAli) [inline python/ray/worker.py:627]: So if “ray://“ is absent it does not use client?

Comment (ckw017) [reply python/ray/worker.py:627]: Yes

File: python/ray/tests/BUILD
Comment (ijrsvt) [inline python/ray/tests/BUILD:137]: Can you move this to medium?

File: python/ray/tests/test_ray_init.py
Comment (ijrsvt) [inline python/ray/tests/test_ray_init.py:167]: Should this actually be misspelled?

File: python/ray/client_builder.py
Comment (ijrsvt) [inline python/ray/client_builder.py:143]: What about `kwargs.get("job_config")`?

File: python/ray/worker.py
Comment (pcmoritz) [inline python/ray/worker.py:628]: `start a ray client session` is a bit confusing, we should say `connect to the cluster at the given address using the Ray client`.

File: python/ray/worker.py
Comment (pcmoritz) [inline python/ray/worker.py:721]: Not sure how this is currently working here, but we want to use this opportunity to clean up these arguments:
- For the old way to call ray.init, all the arguments should be supported as they used to be
- For the new protocol:// convention, we should restrict the parameters that can be passed in. In particular:
For local mode, only
```
ray.init(
    address="local://",
    runtime_env={...},
    num_cpus=5,
    num_gpus=1,
    resources={...},
    dashboard_host="127.0.0.1",
    dashboard_port=8000,
    logging_format=”...”,
    logging_level=20,
    logging_to_driver=False,
    internal_config={...}
)
```
are supported. For ray:// mode, only
```
ray.init(
    address="ray://54.34.99.123:6379",
    namespace="my-namespace",
    job_name="job-0",
    runtime_env={...},
)
```
are supported.

Comment (pcmoritz) [reply python/ray/worker.py:721]: If any extra parameters are specified, an error should be raised. Also we should make sure only the new parameters are showing up in the documentation.

**Comment by jiaodong** [comment]
Hi all, i got pointed to this PR as fix needed to resolve a failing long running test on weekly for 1.5.0 release: 

Buildkite job:
https://buildkite.com/ray-project/periodic-ci/builds/548#de9691e9-3027-44a1-a109-e2ab6e437fd2

Anyscale job failure message:
https://beta.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_419mwHYMfuuL6KqmK4ukG1sy

With error message 

```
Traceback (most recent call last):
  File "distributed/test_distributed.py", line 143, in <module>
    ray.client().env({}).connect()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/client_builder.py", line 97, in connect
    self.address, job_config=self._job_config)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/util/client_connect.py", line 36, in connect
    ignore_version=ignore_version)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/util/client/__init__.py", line 81, in connect
    connection_retries=connection_retries)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/util/client/worker.py", line 156, in __init__
    raise ConnectionError("ray client connection timeout")
ConnectionError: ray client connection timeout
```

Would you please help to confirm:
1) If this issue is the root cause, and should be release blocking 
    - If so, what's our ETA for the fix so I can cherrypick ?
2) If this issue is unrelated or failing test should not be release blocking 

cc: @simon-mo

**Comment by ckw017** [comment]
@jiaodong Not sure if this is a fix, but should make it easier to specify whether you want to:
- Start a local cluster and connect to it with client (ray.init("local://..."))
- Connect to an existing cluster with client (ray.init("ray://..."))
- Connect directly with a driver (use ray.init as before)

Although as far as I know this shouldn't be blocking potential solutions. If you know more details about the issue (using client when you want to connect directly with a driver?, trying to start+connect to a local cluster but attempting to connect to an existing cluster instead) let me know

ETA should be by EOD tomorrow

File: python/ray/worker.py
Comment (ckw017) [reply python/ray/worker.py]: Sounds good. Currently this works by passing everything* the user passed to init into either:
- builder._catch_all, for **kwargs
- builder._store_args, for all the other arguments

This should let us do argument validation differently for each protocol by overriding the methods in the corresponding builder (local:// [can](https://github.com/ckw017/ray/blob/9759246c8de0824e24d3d9fce96ece4e32d3b915/python/ray/tests/test_ray_init.py#L149-L151) take all the regular init args, ray:// [can't](https://github.com/ckw017/ray/blob/9759246c8de0824e24d3d9fce96ece4e32d3b915/python/ray/tests/test_ray_init.py#L171-L176))


* Currently there's a bit of a caveat: since all the init arguments come with defaults, currently figuring out which ones have changed by [comparing everything to their default value](https://github.com/ckw017/ray/blob/9759246c8de0824e24d3d9fce96ece4e32d3b915/python/ray/worker.py#L728-L743), so technically someone might do ray.init("ray://1.2.3.4", num_cpus=None) and won't get an error even though num_cpus isn't a valid argument, since num_cpus is still the default value. But case of people trying to override with default values shouldn't be too prominent or consequential.

**Comment by jiaodong** [comment]
> @jiaodong Not sure if this is a fix, but should make it easier to specify whether you want to:
> 
> * Start a local cluster and connect to it with client (ray.init("local://..."))
> * Connect to an existing cluster with client (ray.init("ray://..."))
> * Connect directly with a driver (use ray.init as before)
> 
> Although as far as I know this shouldn't be blocking potential solutions. If you know more details about the issue (using client when you want to connect directly with a driver?, trying to start+connect to a local cluster but attempting to connect to an existing cluster instead) let me know
> 
> ETA should be by EOD tomorrow

I got to this PR with help from @rkooo567  and @ijrsvt so they might know better :) I'm simply trying to see what's the next action should be to address this failed release test and looks like you're already on it, so I will ping you later tomorrow to see if you need anything from me and i can cherry pick on top of current 1.5.0 / 1.5.0rc0 branch. Thanks !

File: python/ray/tests/test_ray_init.py
Comment (ijrsvt) [inline python/ray/tests/test_ray_init.py:214]: Can we add one more test that looks like the following:

```
@pytest.mark.parametrize("input", [None, "auto"])
def test_ray_address(input, call_ray_start):
    address = call_ray_start
    with unittest.mock.patch.dict(os.environ, {"RAY_ADDRESS": address}):
        res = ray.init(input)
        # Ensure this is not a client.connect()
        assert not isinstance(res, ClientContext)
        ray.shutdown()

    with unittest.mock.patch.dict(os.environ, {"RAY_ADDRESS": "localhost:{}".format(address.split(":")[-1])}):
        res = ray.init(input, ignore_reinit_error=False)
        # Ensure this is not a client.connect()
        assert not isinstance(res, ClientContext)


```

**Comment by ijrsvt** [comment]
@ckw017 @jiaodong 

From what I can tell the issue with the release tests is that they are running in an environment where `RAY_ADDRESS=localhost:6379` (redis) is set and that is causing `client.connect()` to try to Ray Client connect to a Redis address.

File: python/ray/tests/test_ray_init.py
Comment (ijrsvt) [reply python/ray/tests/test_ray_init.py:214]: @rkooo567 @jiaodong I think this test should cover your use case.

**Comment by jiaodong** [comment]
> @ckw017 @jiaodong
> 
> From what I can tell the issue with the release tests is that they are running in an environment where `RAY_ADDRESS=localhost:6379` (redis) is set and that is causing `client.connect()` to try to Ray Client connect to a Redis address.

https://gist.github.com/jiaodong/19cad5b27eaec7badde6dee97d414ed1

This is what i can see from head startup log, essentially 

`ssh: connect to host 52.38.18.250 port 22: Connection timed out`

And from test_distributed.py it's using ray client to start a cluster

`ray.client().env({}).connect()`

Did this ring a bell and have we ever ran into port 22 timeout issue with ray client ?

**Comment by ijrsvt** [comment]
@jiaodong 

> ssh: connect to host 52.38.18.250 port 22: Connection timed out
This is just cluster launcher. Did it eventually succeed? The reason for this error is that we do the AWS API call to launch the node and then immediately try to SSH. There is usually some delay until the node is ready for SSH (hence these failures on trying to execute `uptime`).

**Comment by jiaodong** [comment]
> @jiaodong
> 
> > ssh: connect to host 52.38.18.250 port 22: Connection timed out
> 
> This is just cluster launcher. Did it eventually succeed? The reason for this error is that we do the AWS API call to launch the node and then immediately try to SSH. There is usually some delay until the node is ready for SSH (hence these failures on trying to execute `uptime`).

@ijrsvt for the example I looked into, it didn't, but for core-nightly tests' single_node and object_store failures they succeeded with same error so I think what i discovered should be irrelevant.

So some n00b questions ... is cherry-picking this PR once it's merged the best way to fix all three ray.client() connect timeout errors we saw automatically, or we need to make more changes to existing test code to use different RAY_ADDRESS ?

**Review by ijrsvt — APPROVED** [review]
This looks good from my standpoint. Up to @pcmoritz for the final approval.

**Review by pcmoritz — APPROVED** [review]
LGTM! Next we should clean up the documentation and make sure it shows the new arguments for ray.init only and a link that links to the old arguments. This will likely need some wrangling with sphinx.

**Comment by ckw017** [comment]
Windows failures (test_object_spilling_2, test_multi_tenancy, test_list_actors, test_global_state) all look unrelated a la https://flakey-tests.ray.io/, reviewers can merge at their discretion!

File: doc/source/cluster/ray-client.rst
Comment (richardliaw) [inline doc/source/cluster/ray-client.rst:20]: do you still need `.connect()` here?

File: doc/source/cluster/ray-client.rst
Comment (richardliaw) [inline doc/source/cluster/ray-client.rst:50]: Can we get rid of this? Let's just reference an older version of the docs instead,

**Comment by AmeerHajAli** [comment]
Agree with @richardliaw. @ckw017, can you adress that in a followup PR?

File: doc/source/cluster/ray-client.rst
Comment (ckw017) [reply doc/source/cluster/ray-client.rst:20]: Woops, no, not needed. Will fix in followup

**Pull request merged**
